File: app/src/main/java/com/sgpublic/bilidownload/util/MyLog.java
package com.sgpublic.bilidownload.util;

import android.util.Log;

import com.sgpublic.bilidownload.BuildConfig;

import java.util.HashMap;
import java.util.Map;

public class MyLog {
    private final static boolean out = BuildConfig.DEBUG;

    public static void v(Object message){
        if (out){
            doLog(Log::v, message);
        }
    }

    public static void v(Object message, Throwable e){
        if (out){
            doLog((tag, message1, e1) -> Log.v(tag, message1), message, e);
        }
    }

    public static void d(Object message){
        if (out){
            doLog(Log::d, message);
        }
    }

    public static void d(Object message, Throwable e){
        if (out){
            doLog((tag, message1, e1) -> Log.d(tag, message1), message, e);
        }
    }

    public static void i(Object message){
        if (out){
            doLog(Log::w, message);
        }
    }

    public static void i(Object message, Throwable e){
        if (out){
            doLog((tag, message1, e1) -> Log.i(tag, message1), message, e);
        }
    }

    public static void w(Object message){
        if (out){
            doLog(Log::w, message);
        }
    }

    public static void w(Object message, Throwable e){
        if (out){
            doLog((tag, message1, e1) -> Log.w(tag, message1), message, e);
        }
    }

    public static void e(Object message){
        if (out){
            doLog(Log::e, message);
        }
    }

    public static void e(Object message, Throwable e){
        if (out){
            doLog((tag, message1, e1) -> Log.e(tag, message1), message, e);
        }
    }

    private static void doLog(DoLogSimplify doLog, Object message){
        StackTraceElement ste = new Throwable().getStackTrace()[2];
        String tag_name = "MyLog (" + ste.getFileName() + ":" + ste.getLineNumber() + ")";
        String message_string = String.valueOf(message);
        if (message_string.length() > 1024){
            int index;
            for (index = 0; index < message_string.length() - 1024; index = index + 1024){
                String out = message_string.substring(index, index + 1024);
                doLog.onLog(tag_name, out);
            }
            doLog.onLog(tag_name, message_string.substring(index));
        } else {
            doLog.onLog(tag_name, message_string);
        }
    }

    private static void doLog(DoLog doLog, Object message, Throwable e){
        StackTraceElement ste = new Throwable().getStackTrace()[2];
        String tag_name = "MyLog (" + ste.getFileName() + ":" + ste.getLineNumber() + ")";
        String message_string = String.valueOf(message);
        if (message_string.length() > 1024){
            int index;
            for (index = 0; index < message_string.length() - 1024; index = index + 1024){
                String out = message_string.substring(index, index + 1024);
                doLog.onLog(tag_name, out, e);
            }
            doLog.onLog(tag_name, message_string.substring(index), e);
        } else {
            doLog.onLog(tag_name, message_string + "，[" + e.getClass().getCanonicalName() + "] " + e.getLocalizedMessage(), e);
        }
    }

    private interface DoLogSimplify {
        void onLog(String tag, String message);
    }

    private interface DoLog {
        void onLog(String tag, String message, Throwable e);
    }
}

File: app/src/main/java/io/github/sgpublic/bilidownload/util/MyLog.kt
package io.github.sgpublic.bilidownload.util

import android.util.Log
import io.github.sgpublic.bilidownload.BuildConfig

object MyLog {
    fun v(msg: Any, stack: Int = 0) {
        if (!BuildConfig.DEBUG) {
            return
        }
        doLog(msg, stack) { tag, message ->
            Log.v(tag, message)
        }
    }

    fun v(msg: Any, e: Throwable, stack: Int = 0) {
        if (!BuildConfig.DEBUG) {
            return
        }
        doLog(msg, e, stack) { tag, message ->
            Log.v(tag, message, e)
        }
    }

    fun d(msg: Any, stack: Int = 0) {
        if (!BuildConfig.DEBUG) {
            return
        }
        doLog(msg, stack) { tag, message ->
            Log.d(tag, message)
        }
    }

    fun d(msg: Any, e: Throwable, stack: Int = 0) {
        if (!BuildConfig.DEBUG) {
            return
        }
        doLog(msg, e, stack) { tag, message ->
            Log.d(tag, message, e)
        }
    }

    fun i(msg: Any, stack: Int = 0) {
        if (!BuildConfig.DEBUG) {
            return
        }
        doLog(msg, stack) { tag, message ->
            Log.i(tag, message)
        }
    }

    fun i(msg: Any, e: Throwable, stack: Int = 0) {
        doLog(msg, e, stack) { tag, message ->
            Log.i(tag, message, e)
        }
    }

    fun w(msg: Any, stack: Int = 0) {
        doLog(msg, stack) { tag, message ->
            Log.w(tag, message)
        }
    }

    fun w(msg: Any, e: Throwable, stack: Int = 0) {
        doLog(msg, e, stack) { tag, message ->
            Log.w(tag, message, e)
        }
    }

    fun e(msg: Any, stack: Int = 0) {
        doLog(msg, stack) { tag, message ->
            Log.e(tag, message)
        }
    }

    fun e(msg: Any, e: Throwable, stack: Int = 0) {
        doLog(msg, stack) { tag, message ->
            Log.e(tag, message, e)
        }
    }

    private const val maxLength = 2048
    private fun doLog(message: Any, stack: Int, log: (String, String) -> Unit) {
        val ste = Throwable().stackTrace[stack + 3]
        val tag = "MyLog (" + ste.fileName + ":" + ste.lineNumber + ")"
        val msgStr = message.toString()
        if (msgStr.length <= maxLength) {
            log(tag, msgStr)
            return
        }
        var index = 0
        while (index < msgStr.length - maxLength) {
            val out = msgStr.substring(index, index + maxLength)
            log(tag, out)
            index += maxLength
        }
        log(tag, msgStr.substring(index))
    }

    private fun doLog(message: Any, e: Throwable, stack: Int, log: (String, String) -> Unit) {
        doLog("$message，[" + e.javaClass.canonicalName + "] "
                + e.localizedMessage, stack + 1, log)
    }
}